vpc/service/prune_unused_ips.go

```diff
@@ -0,0 +1,109 @@
+package service
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/Netflix/titus-executor/logger"
+	"github.com/Netflix/titus-executor/vpc/tracehelpers"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+)
+
+func (vpcService *vpcService) getVpcIDs(ctx context.Context) ([]keyedItem, error) {
+	tx, err := vpcService.db.BeginTx(ctx, &sql.TxOptions{
+		ReadOnly: true,
+	})
+	if err != nil {
+		logger.G(ctx).WithError(err).Error("Could not start database transaction")
+		return nil, err
+	}
+	defer func() {
+		_ = tx.Rollback()
+	}()
+	rows, err := tx.QueryContext(ctx, "SELECT vpc_id FROM ip_last_used_v3 GROUP BY vpc_id;")
+	if err != nil {
+		return nil, err
+	}
+
+	ret := []keyedItem{}
+	for rows.Next() {
+		var vid vpcID
+		err = rows.Scan(&vid)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, &vid)
+	}
+	_ = tx.Commit()
+	return ret, nil
+}
+
+type vpcID struct {
+	vpcID string
+}
+
+func (v *vpcID) key() string {
+	return fmt.Sprintf(v.vpcID)
+}
+
+func (v *vpcID) String() string {
+	return fmt.Sprintf(v.vpcID)
+}
+
+func (vpcService *vpcService) pruneLastUsedIPAddresses(ctx context.Context, vpcID keyedItem, tx *sql.Tx) (retErr error) {
+
+	//  Each Titus task requires an IPv4 address. Each VPC has 3 x /16s for Titus to operate in.
+	// That means that the maximum number of "interesting" IPv4 rows per table are 3*(2**16) = 196608
+	howManyToKeep := 200000
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	ctx, span := trace.StartSpan(ctx, "pruneLastUsedIPAddresses")
+	defer span.End()
+
+	logger.G(ctx).WithFields(map[string]interface{}{
+		"vpcID": vpcID.String,
+	}).Info("Beginning purge of last used ip addresses")
+
+	result, err := tx.ExecContext(ctx,
+		`WITH addresses_by_family_in_vpc AS
+		(SELECT id,
+				last_seen,
+				RANK() OVER (PARTITION BY family(ip_address)
+							 ORDER BY last_seen DESC) AS rank
+		FROM ip_last_used_v3 WHERE AND vpc_id = '$2;)
+	    DELETE
+	    FROM ip_last_used_v3
+	    WHERE id IN
+		  (SELECT id
+		   FROM addresses_by_family_in_vpc
+		   WHERE rank > $1 AND last_seen > now() - INTERVAL '30 DAY')
+		AND vpc_id = '$2;`,
+		howManyToKeep,
+		vpcID.String,
+	)
+	if err != nil {
+		err = errors.Wrapf(err, "Failed to delete unused ip adddress for vpc %s", vpcID.String())
+		return err
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		err = errors.Wrap(err, "Failed to load the result")
+		return err
+	}
+
+	logger.G(ctx).WithFields(map[string]interface{}{
```

Comment:
If rows > 0; you may want to add an analyze statement here.

Reply:
Why though? Wouldn't it be pretty normal to remove rows?

Reply:
Eh, I guess we can look at it later.